File: thinking-categorically/applicative-functor/applicative-functor-laws/homomorphism-law.ts
function main() {
  console.log("=== Applicative Homomorphism Law ===\n");

  const result = listHomomorphismExample();
  console.log(`Array<T> result: ${result.join(", ")}`); // 6
}

// --------------------------------
// 1 Array<T> example
// --------------------------------
function listHomomorphismExample(): number[] {
  const times2 = (x: number): number => x * 2;

  // Left: applying array of functions to array of values
  const left = [times2].flatMap(f => [3].map(f));

  // Right: applying function directly to value, then wrapping in array
  const right = [times2(3)];

  console.log(`Left == Right? ${arraysEqual(left, right)}`); // true
  return left;
}

// Helper to check shallow array equality
function arraysEqual<T>(a: T[], b: T[]): boolean {
  return a.length === b.length && a.every((val, idx) => val === b[idx]);
}

main();

File: thinking-categorically/applicative-functor/applicative-functor-laws/homomorphism-law.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

class Program
{
    static async Task Main()
    {
        Console.WriteLine("=== Applicative Homomorphism Law ===\n");

        // 1. Task<T>
        var result1 = await TaskHomomorphismExample();
        Console.WriteLine($"Task<T> result: {result1}"); // 3

        // 2. List<T>
        var result2 = ListHomomorphismExample();
        Console.WriteLine($"List<T> result: {string.Join(", ", result2)}"); // 6
    }

    // --------------------------------
    // 1. Task<T> example (like Maybe)
    // --------------------------------
    static async Task<int> TaskHomomorphismExample()
    {
        Func<int, int> add1 = x => x + 1;
        var left = await Ap(Task.FromResult(add1), Task.FromResult(2));
        var right = await Task.FromResult(add1(2));

        Console.WriteLine($"Left == Right? {left == right}"); // True
        return left;
    }

    static async Task<TResult> Ap<T, TResult>(
        Task<Func<T, TResult>> tf,
        Task<T> tx)
    {
        var f = await tf;
        var x = await tx;
        return f(x);
    }

    // --------------------------------
    // 2. List<T> example
    // --------------------------------
    static List<int> ListHomomorphismExample()
    {
        Func<int, int> times2 = x => x * 2;

        var left = Ap(new List<Func<int, int>> { times2 }, new List<int> { 3 });
        var right = new List<int> { times2(3) };

        Console.WriteLine($"Left == Right? {left.SequenceEqual(right)}"); // True
        return left;
    }

    static List<TResult> Ap<T, TResult>(
        List<Func<T, TResult>> fs,
        List<T> xs)
    {
        return fs.SelectMany(f => xs.Select(f)).ToList();
    }
}
